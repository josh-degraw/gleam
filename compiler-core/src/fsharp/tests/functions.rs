use crate::assert_fsharp;

#[test]
fn function_as_value() {
    assert_fsharp!(
        r#"
fn other() {
  Nil
}

pub fn foo() {
  other
}
"#
    );
}

#[test]
fn test_function_declaration_and_usage() {
    assert_fsharp!(
        "
fn add(a, b) {a + b}

pub fn foo() {
  add(1,2)
}
"
    );
}

#[test]
fn entry_point_main_function_with_no_args() {
    assert_fsharp!(
        "
pub fn main() {
    0
}
"
    );
}
#[test]
fn entry_point_main_function_with_args() {
    assert_fsharp!(
        "
pub fn main(args) {
    0
}
"
    );
}

#[test]
fn function_with_use_statement() {
    assert_fsharp!(
        r#"
fn go() {
  use a <- something
  a + 1
}
fn something(cba) { cba(1) }
"#
    );
}

#[test]
fn function_with_use_statements() {
    assert_fsharp!(
        r#"
fn go() {
  use a <- something
  use b <- something
  a + 1 + b
}
fn something(cba) { cba(1) }
"#
    );
}

#[test]
fn test_function_with_nested_function() {
    assert_fsharp!(
        r#"
type Option(a) {
  Some(a)
  None
}
fn go(a) {
    let inc_or_zero = fn(x) {
        case x {
            Some(i) -> i + 1
            None -> 0
        }
    }
    inc_or_zero(a)
}
"#
    );
}
