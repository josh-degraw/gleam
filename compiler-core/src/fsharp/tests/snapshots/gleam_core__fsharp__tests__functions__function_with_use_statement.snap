---
source: compiler-core/src/fsharp/tests/functions.rs
assertion_line: 54
expression: "\nfn go() {\n  use a <- something\n  a + 1\n}\nfn something(cba) { cba(1) }\n"
---
module rec my.``mod``
let private something (cba: int -> 'u12) = begin
    cba (1)
end
let private go () = begin
    something (fun(a: int) -> begin a + 1 end)
end
